go-controller/pkg/util/dns.go

```diff
@@ -0,0 +1,265 @@
+	if err != nil || config == nil {
+		return nil, fmt.Errorf("cannot initialize the resolver: %v", err)
+	}
+
+	return &DNS{
+		dnsMap:      map[string]dnsValue{},
+		nameservers: filterIPv4Servers(config.Servers),
+		port:        config.Port,
+	}, nil
+}
+
+func (d *DNS) Size() int {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return len(d.dnsMap)
+}
+
+func (d *DNS) Get(dns string) dnsValue {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data
+}
+
+func (d *DNS) GetIPs(dns string) []net.IP {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data.ips
+}
+
+func (d *DNS) Add(dns string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	d.dnsMap[dns] = dnsValue{}
+	_, err := d.updateOne(dns)
+	if err != nil {
+		delete(d.dnsMap, dns)
+	}
+
+	return err
+}
+
+func (d *DNS) Delete(dns string) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	delete(d.dnsMap, dns)
+}
+
+func (d *DNS) Update(dnsName string) (bool, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return d.updateOne(dnsName)
+}
+
+func (d *DNS) updateOne(dns string) (bool, error) {
+	res, ok := d.dnsMap[dns]
+	if !ok {
+		// Should not happen, all operations on dnsMap are synchronized by d.lock
+		return false, fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
+	}
+
+	ips, ttl, err := d.getIPsAndMinTTL(dns)
+	if err != nil {
+		res.nextQueryTime = time.Now().Add(defaultTTL)
+		d.dnsMap[dns] = res
+		return false, err
+	}
+
+	changed := false
+	if !ipsEqual(res.ips, ips) {
+		changed = true
+	}
+	res.ips = ips
+	res.ttl = ttl
+	res.nextQueryTime = time.Now().Add(res.ttl)
+	d.dnsMap[dns] = res
+	return changed, nil
+}
+
+func (d *DNS) getIPsAndMinTTL(domain string) ([]net.IP, time.Duration, error) {
+	ips := []net.IP{}
+	ttlSet := false
+	var ttlSeconds uint32
+
+	for _, server := range d.nameservers {
+		msg := new(dns.Msg)
+		msg.SetQuestion(dns.Fqdn(domain), dns.TypeA)
+
+		dialServer := server
+		if _, _, err := net.SplitHostPort(server); err != nil {
+			dialServer = net.JoinHostPort(server, d.port)
+		}
+		c := new(dns.Client)
+		c.Timeout = 5 * time.Second
+		in, _, err := c.Exchange(msg, dialServer)
+		if err != nil {
+			return nil, defaultTTL, err
+		}
+		if in != nil && in.Rcode != dns.RcodeSuccess {
+			return nil, defaultTTL, fmt.Errorf("failed to get a valid answer: %v", in)
+		}
+
+		if in != nil && len(in.Answer) > 0 {
+			for _, a := range in.Answer {
+				if !ttlSet || a.Header().Ttl < ttlSeconds {
+					ttlSeconds = a.Header().Ttl
+					ttlSet = true
+				}
+
+				switch t := a.(type) {
+				case *dns.A:
+					ips = append(ips, t.A)
+				}
+			}
+		}
+	}
```

Comment:
You need to sort the list of IPs here, since ordering isn't actually guaranteed, and if the order changes, you'll do needless updates.

Reply:
sorted the list of IPs